### ISSUE TITLE
Erroneous data format for unserializing 'SAML2\XML\saml\NameID'

### ISSUE THREAD

**Issue by rwijnen-um** [simplesamlphp/simplesamlphp#1559]
**Describe the bug**
The following error is raised when using version 1.19.4 when trying to login using the package;
_Warning: Erroneous data format for unserializing 'SAML2\XML\saml\NameID' in SimpleSAML\Session->unserialize() (line 223 of <path>/vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/Session.php).

Notice: unserialize(): Error at offset 2288 of 2856 bytes in SimpleSAML\Session->unserialize() (line 223 of <path>/vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/Session.php)._

I noticed that there are NUL characters in that part of the data for private/protected properties;
_O:21:"SAML2\XML\saml\NameID":7:{s:11:"\0*\0nodeName";s:11:"saml:NameID";s:9:"\0*\0Format";s:51:"urn:oasis:names:tc:SAML:2.0:nameid-format:transient";s:15:"\0*\0SPProvidedID";N;s:8:"\0*\0value";s:19:"xxxxxxxx@xxxxxxx.xx";s:16:"\0*\0NameQualifier";s:16:"xxxxxxxxxxxxxxxx";s:18:"\0*\0SPNameQualifier";s:61:"https://xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx.xx/adfs/services/trust";s:10:"\0*\0element";O:10:"DOMElement":0:{}}_
Note; To make it readable, I've converted all NUL characters to "\0".

**To Reproduce**
We use the package 'drupal/simplesamlphp_auth' (version 3.2.0) in a Drupal 9 environment. This package requires the package 'simplesamlphp/simplesamlphp'. After installation, try to login using simplesaml. It is keeping redirected as the unserialize fails. When canceling the redirect-requests, go to any other page and you will see a similar message as indicated above.

**Expected behavior**
Login is correct and no redirection-loops occurs.

**Additional context**
Something must be changed in version 1.19.4 (or any dependency) as changing back to version 1.19.3 fixes the issue.

**Comment by tvdijen** [comment]
Must be related to #1545 then...
What version of PHP are you running?

**Comment by Jeanvanberlo** [comment]
We are running:
PHP 7.4.21 (cli) (built: Jul  1 2021 16:09:23) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.21, Copyright (c), by Zend Technologies

**Comment by tvdijen** [comment]
Are you able to manually update the `simplesamlp/saml2` package and see if that solves the problem?

**Comment by rwijnen-um** [comment]
We installed commit c55ebdf61859c979945e600bf2c94ef3b6cdebf3 of simplesamlphp/saml2 but the error still exists.

**Comment by ruuds** [comment]
Also stumbled onto this. 1.19.3 works for me. It possibly has something to do with the new unserialize-method in simplesamlphp/saml2? See the diffs:

https://github.com/simplesamlphp/simplesamlphp/compare/v1.19.3...v1.19.4
https://github.com/simplesamlphp/saml2/compare/v4.2.3...v4.2.5

**Comment by tvdijen** [comment]
Yes, that's the change I was referring to..  The problem is that v1.19.3 gives errors on PHP 8.1

**Comment by tvdijen** [comment]
@rwijnen-um Is it possible that you have been reusing the same session as before?  Because I don't see how you can still end up with a `DOMElement` in there.. Has the notice changed at all?

**Comment by Jeanvanberlo** [comment]
We made sure that the session was cleared. The same characters are still present. The notice is still the same
Erroneous data format for unserializing 'SAML2\XML\saml\NameID' in SimpleSAML\Session->unserialize() (line 223 of /vendor/simplesamlphp/simplesamlphp/lib/SimpleSAML/Session.php)
Can we do anything to help with debugging?

**Comment by tvdijen** [comment]
I'm not sure.. I wrote a test-script, but I can't reproduce this at all...

**Comment by Jeanvanberlo** [comment]
@tvdijen we noticed that the serialized data contains an "authData" array. The "authData" array contains a "default-sp" array, within this array a key "saml:sp:NameID" (authData->default-sp->saml:sp:NameID) with a resource as data. It seems this is the cause of the unserialize error. When we remove the resource and f.e. replace this with an empty string, the data unserializes properly. Hope this helps.

**Comment by tvdijen** [comment]
Thanks @Jeanvanberlo ! I understand you're using SimpleSAMLphp as a Service Provider?

**Comment by Jeanvanberlo** [comment]
Indeed, we use simplesamlphp as SP for single signon in drupal with our IdP being our institutional adfs.

**Comment by tvdijen** [comment]
The strange thing is... The resource you are witnessing is the DOMElement that I removed in saml2 v4.2.7... I still get the feeling you're looking at a pre-existing session

**Comment by Jeanvanberlo** [comment]
@tvdijen we truncated our kvstore table. And attempted again in an incognito window. Could the session still be in tact then?

**Comment by tvdijen** [comment]
Hmm, table? Are you using SQL as a session store? Maybe that's were we should be looking then..

**Comment by ruuds** [comment]
We are also using the SQL session store. Truncating the table didn't help though.

**Comment by tvdijen** [comment]
Any chance you're using `opcache` and didn't restart php-fpm/apache/nginx after updating the saml2-library?

**Comment by rwijnen-um** [comment]
It is somehow caused by the fact that BaseIDType is implementing Serializable. Serializable is also deprecated as of PHP 8.1 and should therefor not be used, see https://php.watch/versions/8.1/serializable-deprecated

We found a working fix for us which I think is also working for PHP 8.1 but we couldn't test it. Please note that this was just a test and not production-ready, but I guess it can give a direction where to go to actually fix the issue.

We changed the following code in BaseIDType;
Removed the use statement for Serializable at line 17 and removed the implements statement for Serializable at line 19;
`use Serializable;

abstract class BaseIDType`

Changed the serialize method to its magic method;
`public function __serialize(): string`

Changed the unserialize method to its magic method;
`public function __unserialize(array $serialized): void`
Note that the parameter is of type array now, instead of string.

Changed the line
`$doc = DOMDocumentFactory::fromString($serialized);`
to
`$NameQualifier = $serialized["\0*\0NameQualifier"];
$SPNameQualifier = $serialized["\0*\0SPNameQualifier"];
$doc = DOMDocumentFactory::fromString('<root NameQualifier="' . $NameQualifier . '" SPNameQualifier="' . $SPNameQualifier . '"/>');`
in the new unserialize magic method.

**Comment by rwijnen-um** [comment]
When the class implements Serializable, the serialize() and unserialize() methods are not called at all in PHP 7.4. Some magic is happening when you implement Serializable and that magic is breaking it in PHP 7.4.

**Comment by tvdijen** [comment]
I don't see how your example can work, because the signature for  serialize is:

```
public __serialize(): **array**
```

Also, the magic methods were introduced in PHP 7.4, and we still support PHP 7.1, so that's an issue

**Comment by rwijnen-um** [comment]
Than the __serialize() should be changed to. But I thought that serialize method was not called at all anywhere when I debugged some stuff so I did not dive into it very much. I guess that's why no error occured. Because of the NUL characters in the serialized DOMDocument I think it is serialized by PHP itself. Please note that these NUL characters already exist in pervious version of simplesamlphp/saml2 and it was not causing any trouble there, so that's not the reason this fails.

Can I ask why PHP 7.1 is still supported, as it is already end of life for some time? Also; as far as I know adding these magic methods will not break PHP versions before 7.4, as these magic methods are just normal methods there. They will not be called in PHP < 7.4 but I think that's okay because the old code (without special serialize/unserialize methods) was working fine for that versions. Just removing the magic methods is also working in PHP 7.4. Only PHP 8.1 is requiring some additional serialize/unserialize.

**Comment by tvdijen** [comment]
> Can I ask why PHP 7.1 is still supported, as it is already end of life for some time?

That argument doesn't really hold as long as the large OS-vendors keep supporting and updating it.
For example, Red hat 7.9 comes with PHP 5.4 and still has maintenance support for many years to come and they _will_ backport any security-related issues.

**Comment by tvdijen** [comment]
I think saml2 v4.2.8 should fix your problem.. I have to verify that the issue described in #1545 is also still fixed with this change

**Comment by tvdijen** [comment]
@ruuds What's your use-case? Are you also using Drupal? Or stand-alone SSP as a service provider?

**Comment by tvdijen** [comment]
Is anyone able to share a var_dump() of the `authData` array.. Anomynize it as you please, as long as the data-types remain intact..   We've ruled out the SQL backend as a probable cause and our suspicions go towards the drupal plugin, _especially_ because @Jeanvanberlo mentioned a resource in the `authData` that we cannot explain.

**Comment by ruuds** [comment]
Yes, we also use the drupal simplesamlphp_auth module as a SP, but on the other end we also use it as a standalone IDP (with a custom AuthSource), but the error occures at the SP side. I wont be able to supply dumps till tuesday..

**Comment by tvdijen** [comment]
Yes, this is an SP related issue, not IDP.. I have confirmation that saml2 v4.2.8 still works for PHP 8.1, so we may have a workaround.. I'd still like to figure out the exact problem though

**Comment by rwijnen-um** [comment]
We can confirm that the problems are fixed in v4.2.8.

Many thanks for your effort!

**Comment by rwijnen-um** [comment]
> Is anyone able to share a var_dump() of the `authData` array.. Anomynize it as you please, as long as the data-types remain intact.. We've ruled out the SQL backend as a probable cause and our suspicions go towards the drupal plugin, _especially_ because @Jeanvanberlo mentioned a resource in the `authData` that we cannot explain.

This is an anonimized `authData` array;
`'authData' =>
  array (
    'default-sp' =>
    array (
      'saml:AuthenticatingAuthority' =>
      array (
        0 => 'http://xxxx/adfs/services/trust',
      ),
      'saml:AuthnInstant' => xxx,
      'saml:sp:SessionIndex' => '_9e6536e5-0b71-449e-8918-4b41d3271351',
      'saml:sp:AuthnContext' => 'urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport',
      'saml:sp:prevAuth' =>
      array (
        'id' => '_01cc0682-81d9-49b5-8362-39505e9fb70f',
        'issuer' => 'http://xxxx/adfs/services/trust',
        'inResponseTo' => '_820af0645b0f37898344e80d7fd95c8572c6c4e1a5c',
        'redirect' => 'https://xxx/saml_login',
      ),
      'saml:sp:IdP' => 'http://xxx/adfs/services/trust',
      'Attributes' =>
      array (
        'DisplayName' =>
        array (
          0 => 'xxx',
        ),
        'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress' =>
        array (
          0 => 'xxx@xxx.nl',
        ),
        'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/upn' =>
        array (
          0 => 'xxx@xxx.nl',
        ),
        'http://schemas.microsoft.com/ws/2008/06/identity/claims/groupsid' =>
        array (
          0 => 'xxx',
        ),
        'NameID' =>
        array (
          0 => 'xxx@xxx.nl',
        ),
      ),
      'LogoutState' =>
      array (
        'saml:logout:Type' => 'saml2',
        'saml:logout:IdP' => 'http://xxx/adfs/services/trust',
        'saml:logout:NameID' =>
        __PHP_Incomplete_Class::__set_state(array(
           '__PHP_Incomplete_Class_Name' => 'SAML2\\XML\\saml\\NameID',
           'nodeName' => 'saml:NameID',
           'Format' => 'urn:oasis:names:tc:SAML:2.0:nameid-format:transient',
           'SPProvidedID' => NULL,
           'value' => 'xxx@xxx.nl',
           'NameQualifier' => 'xxx',
           'SPNameQualifier' => 'http://xxx/adfs/services/trust',
           'element' =>
          DOMElement::__set_state(array(
          )),
        )),
        'saml:logout:SessionIndex' => '_9e6436e5-0b71-449e-1218-4b41d32685351',
      ),
      'saml:sp:NameID' => RESOURCE_THAT_BREAKS_UNSERIALIZE,
      'Authority' => 'default-sp',
      'AuthnInstant' => xxx,
      'Expire' => 1642031609,
    ),
  ),
)`

**Comment by ruuds** [comment]
I cannot reproduce the issue anymore, even with the exact same composer.lock files...

**Comment by tvdijen** [comment]
Well, the authData shows that there is/was still a DOMElement in there, that was removed in 4.2.6 already.. Somehow something from an older session must have interfered even though the SQL table was cleared...
Anyway, I'm glad it was solved!
